### PR TITLE
Fix docker build (Work In Progress)

### PR DESCRIPTION
**Pull request by johnoliver**
Creating PR so people can comment on it. 

Not totally happy with this solution. At the moment I am not sure it is worth trying to keep the docker and native builds in the same script, may be cleaner to have separate build scripts and then extract the common parts of each.

I also cannot guarantee that this does not break the native build as the native build did not work for me to start with...will look into the issues with that separately.

Fully or partially fixes issue #64

### CODE REVIEW

File: makejdk.sh [hunk 1]
```diff
@@ -24,17 +24,13 @@
...
-REPOSITORY=AdoptOpenJDK/openjdk-jdk8u
-OPENJDK_REPO_NAME=openjdk
-
-OS_KERNEL_NAME=$(uname | awk '{print tolower($0)}')
-OS_MACHINE_NAME=$(uname -m)
 
-JVM_VARIANT=${JVM_VARIANT:-server}
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+# shellcheck source=sbin/functions.sh
+source "$SCRIPT_DIR/sbin/functions.sh"
```
Comment (karianna) [inline makejdk.sh]: Being very nit-picky here - but perhaps common-functions.sh is a better name?

File: makejdk.sh [hunk 2]
```diff
@@ -24,17 +24,13 @@
...
 
-BUILD_TYPE=normal
-DEFAULT_BUILD_FULL_NAME=${OS_KERNEL_NAME}-${OS_MACHINE_NAME}-${BUILD_TYPE}-${JVM_VARIANT}-release
-BUILD_FULL_NAME=${BUILD_FULL_NAME:-"$DEFAULT_BUILD_FULL_NAME"}
+REPOSITORY=AdoptOpenJDK/openjdk-jdk8u
```
Comment (karianna) [inline makejdk.sh]: I know that @neomatrix369 was going to PR this later as well - but we'll need to deal with repos other than openjdk-jdk8u at some stage.  Don't think you need to do anything yet, but thought I should mention it.

File: makejdk.sh [hunk 3]
```diff
@@ -186,13 +184,14 @@ cloneOpenJDKGitRepo()
 testOpenJDKViaDocker()
 {
   if [[ ! -z $JTREG ]]; then
-    docker run --privileged -t -v "${WORKING_DIR}/${OPENJDK_REPO_NAME}":/openjdk/jdk8u/openjdk --entrypoint jtreg.sh "${CONTAINER}"
+    mkdir -p "${WORKING_DIR}/target"
+    docker run -t -v "${WORKING_DIR}/${OPENJDK_REPO_NAME}":/openjdk/jdk8u/openjdk -v "${WORKING_DIR}/sbin":/openjdk/sbin -v "${WORKING_DIR}/target":/openjdk/target --entrypoint /openjdk/sbin/jtreg.sh "${CONTAINER}"
   fi
```
Comment (karianna) [inline makejdk.sh]: Same note on hard coding of jdk8u, but maybe discuss with @neomatrix369 first.

File: sbin/build.sh [hunk 4]
```diff
@@ -16,20 +16,23 @@
 # Script to download any additional packages for building OpenJDK
 # before calling ./configure (using JDK 7 as the base)
 
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+# shellcheck source=sbin/functions.sh
+source "$SCRIPT_DIR/functions.sh"
```
Comment (karianna) [inline sbin/build.sh]: common-functions.sh?

File: sbin/jtreg.sh [hunk 5]
```diff
@@ -13,6 +13,10 @@
 # limitations under the License.
 #
 
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+# shellcheck source=sbin/functions.sh
+source "$SCRIPT_DIR/functions.sh"
+
```
Comment (karianna) [inline sbin/jtreg.sh]: common-functions.sh?

File: sbin/jtreg.sh [hunk 6]
```diff
@@ -22,12 +26,20 @@ JTREG_VERSION=${JTREG_VERSION:-4.2.0-tip}
 JTREG_TARGET_FOLDER=${JTREG_TARGET_FOLDER:-jtreg}
 JOB_NAME=${JOB_NAME:-OpenJDK}
 NUM_PROCESSORS=${NUM_PROCESSORS:-$(getconf _NPROCESSORS_ONLN)}
+TMP_DIR=/tmp/
```
Comment (karianna) [inline sbin/jtreg.sh]: Linux only path?

File: sbin/jtreg.sh [hunk 7]
```diff
@@ -42,14 +54,15 @@ downloadJtregAndSetupEnvironment()
    echo "Downloading Jtreg binary"
    JTREG_BINARY_FILE="jtreg-${JTREG_VERSION}.tar.gz"
 
-   wget https://ci.adoptopenjdk.net/job/jtreg/lastSuccessfulBuild/artifact/"$JTREG_BINARY_FILE"
+   cd "$TMP_DIR" && wget https://ci.adoptopenjdk.net/job/jtreg/lastSuccessfulBuild/artifact/"$JTREG_BINARY_FILE"
```
Comment (karianna) [inline sbin/jtreg.sh]: Better to split those lines out for debug purposes?

File: sbin/jtreg.sh [hunk 8]
```diff
@@ -42,14 +54,15 @@ downloadJtregAndSetupEnvironment()
...
 
    if [ $? -ne 0 ]; then
      echo "Failed to retrieve the jtreg binary, exiting"
      exit
    fi
 
-   tar xf "$JTREG_BINARY_FILE"
+   
+   cd "$WORKING_DIR" && tar xf "$TMP_DIR/$JTREG_BINARY_FILE"
   fi
```
Comment (karianna) [inline sbin/jtreg.sh]: Better to split those lines out for debug purposes?

File: makejdk.sh [hunk 9]
```diff
@@ -186,13 +184,14 @@ cloneOpenJDKGitRepo()
...
 }
 
 buildAndTestOpenJDKViaDocker()
 {
-  if [ -z "$(which docker)" ] || [ -z "$(pgrep "docker")" ]; then
+  if [ -z "$(which docker)" ]; then
```
Comment (gdams) [inline makejdk.sh]: why did you remove the docker `pgrep` check? this should be fine as it simply checks that docker is running on your machine

File: sbin/build.sh [hunk 10]
```diff
@@ -204,6 +130,7 @@ stepIntoTheWorkingDirectory()
 
 runTheOpenJDKConfigureCommandAndUseThePrebuildConfigParams()
 {
+  cd "$OPEN_JDK_DIR" || exit
```
Comment (gdams) [inline sbin/build.sh]: please change to `"${OPEN_JDK_DIR}"`, or perhaps would `"${OPENJDK_DIR}"` be better here?

File: sbin/build.sh [hunk 11]
```diff
@@ -231,6 +158,8 @@ runTheOpenJDKConfigureCommandAndUseThePrebuildConfigParams()
 
 buildOpenJDK()
 {
+  cd "$OPEN_JDK_DIR" || exit
```
Comment (gdams) [inline sbin/build.sh]: see above

File: sbin/jtreg.sh [hunk 12]
```diff
@@ -66,20 +79,20 @@ downloadJtregAndSetupEnvironment()
 applyingJCovSettingsToMakefileForTests()
 {
   echo "Apply JCov settings to Makefile..." 
-  cd "$WORKING_DIR/$OPENJDK_REPO_NAME/jdk/test" || exit
+  cd "$OPEN_JDK_DIR/jdk/test" || exit
```
Comment (gdams) [inline sbin/jtreg.sh]: see above

File: sbin/jtreg.sh [hunk 13]
```diff
@@ -66,20 +79,20 @@ downloadJtregAndSetupEnvironment()
...
   pwd
 
   sed -i "s/-vmoption:-Xmx512m.*/-vmoption:-Xmx512m -xml:verify -jcov\/classes:$(ABS_PLATFORM_BUILD_ROOT)\/jdk\/classes\/  -jcov\/source:$(ABS_PLATFORM_BUILD_ROOT)\/..\/..\/jdk\/src\/java\/share\/classes  -jcov\/include:*/" Makefile
 
-  cd "$WORKING_DIR/$OPENJDK_REPO_NAME/" || exit
+  cd "$OPEN_JDK_DIR" || exit
```
Comment (gdams) [inline sbin/jtreg.sh]: see above

File: makejdk.sh (hunk 3 above)
Comment (neomatrix369) [inline makejdk.sh]: both @gdams and I can answer about this

Comment (neomatrix369) [inline makejdk.sh]: I would suggest do this change as a separate PR - it might lead to cascade of changes and more testing

File: sbin/jtreg.sh (hunk 6 above)
Comment (neomatrix369) [inline sbin/jtreg.sh]: use `$TMPDIR` (non-windows) and `$TEMP` (Windows)
Will `$TMPDIR` work on other than Linux?

**Comment by neomatrix369** [comment]
Just ran this PR changes via https://ci.adoptopenjdk.net/view/work%20in%20progress/job/openjdk_build_x86-64_linux_johno_test/11/console - looks good :+1:

File: makejdk.sh [hunk 14]
```diff
@@ -24,17 +24,13 @@
...
-JVM_VARIANT=${JVM_VARIANT:-server}
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+# shellcheck source=sbin/functions.sh
+source "$SCRIPT_DIR/sbin/functions.sh"
 
-BUILD_TYPE=normal
-DEFAULT_BUILD_FULL_NAME=${OS_KERNEL_NAME}-${OS_MACHINE_NAME}-${BUILD_TYPE}-${JVM_VARIANT}-release
-BUILD_FULL_NAME=${BUILD_FULL_NAME:-"$DEFAULT_BUILD_FULL_NAME"}
+REPOSITORY=AdoptOpenJDK/openjdk-jdk8u
+OPENJDK_REPO_NAME=openjdk
```
Comment (neomatrix369) [inline makejdk.sh]: could this be an overrideable param as well?
OPENJDK_REPO_NAME=${OPENJDK_REPO_NAME:-openjdk}

**Comment by neomatrix369** [comment]
Just added this other build to test docker work on Jenkins - https://ci.adoptopenjdk.net/view/work%20in%20progress/job/openjdk_build_x86-64_linux_john_docker_test but does not seem to run, needs docker installed on the instance - @gdams worth doing this on the jenkins instance to test our docker build, we should have atleast one such build running to check if our docker build hasn't broken with our native changes

File: sbin/functions.sh [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+#
+
+# Common functions to be used in scripts
+
+
+ALSA_LIB_VERSION=${ALSA_LIB_VERSION:-1.0.27.2}
+FREETYPE_FONT_SHARED_OBJECT_FILENAME=libfreetype.so.6.5.0
+FREETYPE_FONT_VERSION=${FREETYPE_FONT_VERSION:-2.4.0}
+
+determineBuildProperties() {
```
Comment (neomatrix369) [inline sbin/functions.sh]: :+1:

File: makejdk.sh [hunk 16]
```diff
@@ -186,13 +184,14 @@ cloneOpenJDKGitRepo()
 testOpenJDKViaDocker()
 {
   if [[ ! -z $JTREG ]]; then
-    docker run --privileged -t -v "${WORKING_DIR}/${OPENJDK_REPO_NAME}":/openjdk/jdk8u/openjdk --entrypoint jtreg.sh "${CONTAINER}"
+    mkdir -p "${WORKING_DIR}/target"
+    docker run -t -v "${WORKING_DIR}/${OPENJDK_REPO_NAME}":/openjdk/jdk8u/openjdk -v "${WORKING_DIR}/sbin":/openjdk/sbin -v "${WORKING_DIR}/target":/openjdk/target --entrypoint /openjdk/sbin/jtreg.sh "${CONTAINER}"
```
Comment (neomatrix369) [inline makejdk.sh]: In `/openjdk/jdk8u/openjdk`, not clear why we have subfolder `openjdk` inside `/openjdk/jdk8u/` maybe its a good time to resolve this

File: makejdk.sh (hunk 2 above)
Comment (neomatrix369) [inline makejdk.sh]: Clearly `jdk8u` should be generified*, as we will have other variants to come

*not even a word

File: makejdk.sh (hunk 14 above)
Comment (neomatrix369) [inline makejdk.sh]: OPENJDK_REPO_NAME=${OPENJDK_REPO_NAME:-openjdk} (mentioned in another comment)

Comment (neomatrix369) [inline makejdk.sh]: Same for `REPOSITORY`

**Comment by neomatrix369** [comment]
Run the script locally and find it fails:

```
2017-04-24 12:06:51 (761 KB/s) - 'jtreg-4.2.0-tip.tar.gz' saved [7218784/7218784]

List contents of the jtreg folder
ls: cannot access /openjdk/jdk8u//jtreg/*: No such file or directory
Apply JCov settings to Makefile...
/openjdk/jdk8u/openjdk/jdk/test
/openjdk/sbin/jtreg.sh: line 85: ABS_PLATFORM_BUILD_ROOT: command not found
/openjdk/sbin/jtreg.sh: line 85: ABS_PLATFORM_BUILD_ROOT: command not found
Setting up environment variables for JTREG to run
/openjdk/jdk8u//openjdk//build/linux-x86_64-normal-server-release/images/j2sdk-image
ASSEMBLY_EXCEPTION  THIRD_PARTY_README	demo	 jre  man      sample
LICENSE		    bin			include  lib  release  src.zip
Running jtreg via make command (debug logs enabled)
/openjdk/jdk8u/openjdk//make/Main.gmk:32: /home/satyasai/LJCAdvocate/OpenJDKStuff/Tools/openjdk-build/openjdk/make/common/MakeBase.gmk: No such file or directory
make: *** No rule to make target `/home/satyasai/LJCAdvocate/OpenJDKStuff/Tools/openjdk-build/openjdk/make/common/MakeBase.gmk'.  Stop.
Archiving your jtreg results (includes jcov reports)
Package test output into archives...
/openjdk/jdk8u/openjdk
Tarring and zipping the 'testoutput' folder into artefact: OpenJDK-testoutput-with-jcov-reports.tar.gz
Package jcov reports into archives...
/openjdk/jdk8u
/openjdk/jdk8u/openjdk/build/linux-x86_64-normal-server-release
Tarring and zipping the 'testoutput/../jcov' folder into artefact: OpenJDK-jcov-results-only.tar.gz
tar: testoutput/*/JTreport/jcov: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
grep: warning: GREP_OPTIONS is deprecated; please use an alias or script
Copying to the host with docker cp 22c142ce2e30:/openjdk/jdk8u/OpenJDK.tar.gz /home/satyasai/LJCAdvocate/OpenJDKStuff/Tools/openjdk-build
Error response from daemon: lstat /var/lib/docker/aufs/mnt/aca22bc9eea3be617fab732b0be9f2156f16497b30656012d60a008e8a86f71f/openjdk/jdk8u/OpenJDK.tar.gz: no such file or directory
```
Commands use:

```
./makejdk-any-platform.sh
```

and

```
./makejdk-any-platform.sh --destination $(pwd)
```

File: makejdk.sh (hunk 9 above)
Comment (johnoliver) [inline makejdk.sh]: Not sure about others, but for me I have no docker processes running until I first call the docker command, so this entry prevents the docker running unless I have already run the docker command before executing this script

**Comment by neomatrix369** [comment]
@johnoliver have you pushed the `common-functions.sh`? I get the below error when running the scripts:
```
./makejdk.sh: line 30: .../openjdk-build/sbin/common-functions.sh: No such file or directory
./makejdk.sh: line 43: determineBuildProperties: command not found
```

File: sbin/jtreg.sh [hunk 17]
```diff
@@ -22,12 +26,20 @@ JTREG_VERSION=${JTREG_VERSION:-4.2.0-tip}
 JTREG_TARGET_FOLDER=${JTREG_TARGET_FOLDER:-jtreg}
 JOB_NAME=${JOB_NAME:-OpenJDK}
 NUM_PROCESSORS=${NUM_PROCESSORS:-$(getconf _NPROCESSORS_ONLN)}
+TMP_DIR=$(dirname $(mktemp -u))
+OPENJDK_DIR="$WORKING_DIR/$OPENJDK_REPO_NAME"
+TARGET_DIR="$WORKING_DIR"
+
+determineBuildProperties
```
Comment (neomatrix369) [inline sbin/jtreg.sh]: Where is this function defined? In `common-functions.sh` i guess

Comment (neomatrix369) [inline sbin/jtreg.sh]: Still getting this https://github.com/AdoptOpenJDK/openjdk-build/pull/82#issuecomment-296641587, but that could be due to the above things I have pointed out.

File: makejdk.sh (hunk 9 above)
Comment (karianna) [inline makejdk.sh]: I'm the same as @johnoliver I don't always have Docker running first

Comment (gdams) [inline makejdk.sh]: hmm okay, lets remove it then

**Comment by johnoliver** [comment]
I have readded the missing file

**Comment by karianna** [comment]
@johnoliver It's failing some shellcheck on Travis.

**Comment by neomatrix369** [comment]
@johno are you able to execute a native build after building via docker, I mean what happens after you have done these two commands:

Clear everything and start from scratch
```
./makejdk-any-plattform.sh     <=== docker build
```
```
./makejdk-any-plattform.sh    --source <some source> --destination <some destination>
```

Gives me this error: cannot create directory: /openjdk

Clear everything and start from scratch
```
./makejdk-any-plattform.sh    --source <some source> --destination <some destination>
```
```
./makejdk-any-plattform.sh     <=== docker build
```
Works fine no complains.

The above use case might not be common but nevertheless it happens.

I'm thinking there might be some ownership/permission issue with the `openjdk` when the docker build creates it.

Try to bring commits from this repo into yours otherwise it might not work for everyone - https://github.com/neomatrix369/openjdk-build/commits/fix-docker-build

Can someone also please test both native and docker builds for us?

**Comment by neomatrix369** [comment]
Kicked off a build on https://ci.adoptopenjdk.net/view/work%20in%20progress/job/openjdk_build_x86-64_linux_johno_test/12/console to test native build. Docker builds will still need to be tested locally by one of us.

**Comment by neomatrix369** [comment]
Native build is successful, see https://ci.adoptopenjdk.net/view/work%20in%20progress/job/openjdk_build_x86-64_linux_johno_test/13/console, while docker build is (mostly) good, see https://github.com/AdoptOpenJDK/openjdk-build/pull/82#issuecomment-297985119.

To fix the permission issue (cannot create...) we could open a new issue and investigate it separately.

**Comment by karianna** [comment]
@johnoliver - still some shellcheck fixes to go I think.

**Comment by neomatrix369** [comment]
@johnoliver As mentioned on the slack channel, the new changes work good, haven't tried the combination as mentioned [earlier](https://github.com/AdoptOpenJDK/openjdk-build/pull/82#issuecomment-297985119).

**Comment by neomatrix369** [comment]
@karianna @gdams Can we please review this PR, it looks good to go, worth merging now
